### PR TITLE
Fixed #24747 -- Allowed transforms in QuerySet.order_by()

### PR DESCRIPTION
**Pull request by MatthewWilkes**
This is a PR to fix #24747, and other related tickets. It allows order_by to include results of arbitrary transforms.

This works by making setup_joins understand transforms and return an additional value that converts a fiel object to something the SQL compiler can understand that takes the transforms into account.

Practically, this allows order_by and distinct on JSONField, HStoreField and ArrayField.

### CODE REVIEW

File: django/db/models/sql/compiler.py [hunk 1]
```diff
@@ -651,7 +651,9 @@ def find_ordering_name(self, name, opts, alias=None, default_order='ASC',
                                                        order, already_seen))
             return results
         targets, alias, _ = self.query.trim_joins(targets, joins, path)
-        return [(OrderBy(t.get_col(alias), descending=descending), False) for t in targets]
```
Comment (MatthewWilkes) [inline django/db/models/sql/compiler.py]: I would appreciate some guidance here, @jarshwah @charettes (and anyone else..)

It's my understanding that all the members of `targets` and `final_field` are guaranteed to have the same value, so I don't understand why we're looping over targets and generating `OrderBy` objects. It seems to me that if there was a reason we needed to generate multiple `ORDER BY` clauses then the `final_field` should be part of this too. It seems that just using `final_field` and not `targets` produces correct results, but I fear there is a subtlety of joins that I'm missing.

File: django/db/models/sql/query.py [hunk 2]
```diff
@@ -1401,9 +1402,42 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True):
...
+        transforms = []
+        # Try resolving all the names as fields first, but if there is an error
+        # progressively treat trailing names as lookups until we can resolve
+        # a field
+        last_field_exception = None
+        for pivot in range(len(names), 0, -1):
+            try:
+                path, final_field, targets, rest = self.names_to_path(
+                    names[:pivot], opts, allow_many, fail_on_missing=True)
+            except (FieldError) as exc:
```
Comment (pope1ni) [inline django/db/models/sql/query.py]: Drop the unneeded parentheses.

File: django/db/models/sql/query.py [hunk 3]
```diff
@@ -1401,9 +1402,42 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True):
...
+                # resolved into fields
+                transforms = names[pivot:]
+                break
+        for name in transforms:
+            try:
+                if hasattr(final_field, 'resolve_expression'):
+                    final_field = self.try_transform(final_field, name)
+                else:
+                    final_field = self.try_transform(final_field.get_col(alias), name)
+            except (FieldError):
```
Comment (pope1ni) [inline django/db/models/sql/query.py]: Drop the unneeded parentheses.

File: django/db/models/sql/query.py [hunk 4]
```diff
@@ -1401,9 +1402,42 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True):
...
+                # The transforms are the remaining items that couldn't be
+                # resolved into fields
+                transforms = names[pivot:]
+                break
+        for name in transforms:
+            try:
+                if hasattr(final_field, 'resolve_expression'):
+                    final_field = self.try_transform(final_field, name)
+                else:
+                    final_field = self.try_transform(final_field.get_col(alias), name)
```
Comment (pope1ni) [inline django/db/models/sql/query.py]:
```python
if not hasattr(final_field, 'resolve_expression'):
    final_field = final_field.get_col(alias)
try:
    final_field = self.try_transform(final_field, name)
    ...
```

File: django/db/models/sql/query.py (hunk 2 above)
Comment (MatthewWilkes) [inline django/db/models/sql/query.py]: Thank you, but as I said in the description I'm lokking for help with the mewt of the problem right now, not style advice. This does currently contain WIP code that is not suitable for review.

File: django/db/models/sql/query.py (hunk 4 above)
Comment (MatthewWilkes) [inline django/db/models/sql/query.py]: Your suggestion here is actually wrong, it needs to be in a temporary variable, not final_field, otherwise if an exception is raised final_field is left as a Col rather than a Field.

File: django/db/models/sql/compiler.py (hunk 1 above)
Comment (MatthewWilkes) [inline django/db/models/sql/compiler.py]: I've played some more here and I think the code ends up too messy this way. I'm now exploring returning a function as part of setup_joins that returns an object the compiler can handle. For fields this is equivalent to get_col. This should get around the problems around joins I was worried about. I'll push a change once I'm home from pycon.

Comment (MatthewWilkes) [inline django/db/models/sql/compiler.py]: To get this to work, I needed to allow `get_distinct` and `distinct_sql` to return a result, params tuple, rather than just a result, but I think this is a better approach. It's allowed me to get `DISTINCT ON` working with JSON keys. To get that, though, I had to add an `as_text` transform to get access to `#>>`, as JSON keys aren't orderable, unlike JSONB keys.

I'm still playing, and the branch is dirty, but this has got rid of my concerns above, so I think I'm on the right track.

**Comment by MatthewWilkes** [comment]
The next things I'll be doing is adding the tests for HStore, ArrayField and JSONB lookups in postgres, mirroring the JSON ones.  If anyone has advice on other things to be tested, or if they think my approach has flaws I'd very much like to hear. As it stands, I'm happy with the approach of adding a transform function as a return from setup_joins.

**Comment by MatthewWilkes** [comment]
buildbot, test on oracle

**Comment by felixxm** [comment]
buildbot, test on oracle

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: django/db/models/sql/query.py [hunk 5]
```diff
@@ -1401,10 +1403,50 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True):
...
+            return field.get_col(alias)
+
+        # Try resolving all the names as fields first, but if there is an error
+        # progressively treat trailing names as lookups until we can resolve
+        # a field
+        last_field_exception = None
+        for pivot in range(len(names), 0, -1):
+            try:
+                path, final_field, targets, rest = self.names_to_path(
+                    names[:pivot], opts, allow_many, fail_on_missing=True)
```
Comment (jarshwah) [inline django/db/models/sql/query.py]: I think we prefer the closing paren on a newline

File: django/db/models/sql/query.py [hunk 6]
```diff
@@ -1401,10 +1403,50 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True):
...
+                try:
+                    wrapped = previous(field, alias)
+                    return self.try_transform(wrapped, name)
+                except FieldError:
+                    # FieldError is raised if the transform does not exist.
+                    # AttributeError may be raised when formatting the exception
+                    # message, but we're going to reraise an earlier exception
+                    # so we can swallow this.
+                    if isinstance(final_field, Field) and last_field_exception:
+                        raise last_field_exception
```
Comment (jarshwah) [inline django/db/models/sql/query.py]: I'm leaving a comment for me to check later. Does it make sense to raise current exception from last_field_exception? I think that'll show the name not being a transform caused by the name not being a field.

File: django/db/models/sql/query.py [hunk 7]
```diff
@@ -1401,10 +1403,50 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True):
...
+                    raise
+                else:
+                    last_field_exception = exc
+            else:
+                # The transforms are the remaining items that couldn't be
+                # resolved into fields
+                transforms = names[pivot:]
+                break
+        for name in transforms:
+            def transform(field, alias, name, previous):
```
Comment (jarshwah) [inline django/db/models/sql/query.py]: `def transform(field, alias, *, name, previous)` ?

Makes the partial application further down more obvious that name and previous are supplied by keyword args. Not certain this is better, but I had to double check the partial applicability worked correctly.

File: django/db/models/sql/query.py [hunk 8]
```diff
@@ -1418,7 +1460,7 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True):
             reuse = can_reuse if join.m2m else None
             alias = self.join(connection, reuse=reuse)
             joins.append(alias)
-        return final_field, targets, opts, joins, path
+        return final_field, targets, opts, joins, path, final_transformer
```
Comment (jarshwah) [inline django/db/models/sql/query.py]: This is getting out of hand, especially considering callers ignoring so many of the return values. Does it make sense to create a type to model these values? At a minimum I think we should consider namedtuple.

File: tests/custom_lookups/tests.py [hunk 9]
```diff
@@ -63,6 +63,15 @@ def as_sql(self, compiler, connection):
         return '3 * (%s)' % lhs, lhs_params
 
 
+class LastDigitBilateralTransform(models.Transform):
+    bilateral = True
```
Comment (jarshwah) [inline tests/custom_lookups/tests.py]: Does this make a difference to the tests? I'm not a fan of the bilateral attribute - it was introduced before expressions were a thing. Not sure we want to propagate their use.

**Comment by MatthewWilkes** [comment]
> .values('datefield__month') supported?

It is, I'll add an explicit test. `.values` is demonstrated in the postgres tests too, as:

```python
   def test_deep_values(self):
       query = JSONModel.objects.values_list('field__k__l')
       self.assertSequenceEqual(
           query,
           [(None, ), (None, ), (None, ), (None, ), (None, ), (None, ),
            (None, ), (None, ), ('m', ), (None, ), (None, ), (None, )]
       )
```

but I agree it should be in the platform agnostic suite.

> Any tests for multiple transformations? datefield__month__lastdigit or datefield__date__day?

Good idea.

> Any distinct tests?

Yeah, there are some in the postgres suite. I didn't do any in the backend agnostic as it's only really helpful to do it on distinct on, I think?

```python
   def test_deep_distinct(self):
       query = JSONModel.objects.distinct('field__k__l__as_text').values_list('field__k__l')
       self.assertSequenceEqual(
           query,
           [('m', ), (None, )]
       )
```

I added the `as_text` transform that works the same way as the JSON `KeyTransform` in that it merges multiple transforms together to give the right output. Using as_text at the end causes the `->>` operator to be used instead of `->`, which then compares the text representation rather than the data. This is needed as you don't have a general guarantee of having equality operators defined for the JSON field.

I should document that, too.

File: django/db/models/sql/query.py (hunk 8 above)
Comment (MatthewWilkes) [inline django/db/models/sql/query.py]: No argument here, I got a bit of underscore blindness at times. Think that should be its own PR, so we don't have too much refactoring in with functionality changes?

File: tests/custom_lookups/tests.py (hunk 9 above)
Comment (MatthewWilkes) [inline tests/custom_lookups/tests.py]: It doesn't, I only did it to match with `mul3` and `div3`, as that seemed the closest match in the test suite for what I was trying to demonstrate.

File: django/db/models/sql/query.py (hunk 6 above)
Comment (MatthewWilkes) [inline django/db/models/sql/query.py]: Hm, maybe. I think that perhaps we should raise it with a different context depending on if it's the first transform in the chain or not. I suspect that users are more likely to typo a field name or a join than they are a transform, so we'd want to push them to the (very helpful) `django.core.exceptions.FieldError: Cannot resolve keyword 'notafield' into field. Choices are: article, email, first_name, id, last_name` type errors we generate.

That said, if someone is doing ``datetimefield__date__day`` we don't really want to say that `date` isn't a field if they typo `day`.

File: django/db/models/sql/query.py (hunk 8 above)
Comment (timgraham) [inline django/db/models/sql/query.py]: Yes please

**Comment by timgraham** [comment]
It would be nice to update the docs/commit message/ticket, etc. to reflect the fact that `JSONField` `values_list` functionality. If the changes aren't coupled, even a separate commit for that may make review easier.

File: django/contrib/postgres/fields/jsonb.py [hunk 10]
```diff
@@ -118,6 +118,24 @@ class KeyTextTransform(KeyTransform):
     _output_field = TextField()
 
 
+class CoerceToText(Transform):
```
Comment (sir-sigurd) [inline django/contrib/postgres/fields/jsonb.py]: I think that this should be added in the separate pull request.

Comment (charettes) [inline django/contrib/postgres/fields/jsonb.py]: Also, I think this can be achieved by defining `KeyTextTransform.lookup_name` and registering it as a lookup. The reason why this wasn't done in the first place is because of the possible `lookup_name` (e.g. `'as_text'`) and jsonb key collision. See [#27257](https://code.djangoproject.com/ticket/26511) for more details.

Comment (MatthewWilkes) [inline django/contrib/postgres/fields/jsonb.py]: Hm, yeah, these are good pointers. The problem I had that initially made me add it was to make sure the value returned in postgres was fully orderable, as just using the normal `KeyTransform` was causing errors with comparitor operations being undefined. That doesn't really make sense to me, as I thought that was one of the things JSONB was supposed to solve.

Does it make sense to have lookups that convert various things that could be in JSONFields to other Django primitive fields, so we can chain the lookups?

Splitting the PR out makes sense, I've been splitting the `setup_joins` stuff out already, so can do this one too.

**Comment by MatthewWilkes** [comment]
Just want to follow up and apologise for not moving this on for weeks now. I've not forgotten, I'm going to try and find some time this week or next to get all the comments addressed.

File: django/db/models/sql/compiler.py [hunk 11]
```diff
@@ -605,21 +610,22 @@ def get_distinct(self):
...
-            _, targets, alias, joins, path, _ = self._setup_joins(parts, opts, None)
+            field, targets, alias, joins, path, _, transform_function = self._setup_joins(parts, opts, None)
             targets, alias, _ = self.query.trim_joins(targets, joins, path)
             for target in targets:
                 if name in self.query.annotation_select:
                     result.append(name)
                 else:
-                    result.append("%s.%s" % (qn(alias), qn2(target.column)))
-        return result
+                    r, p = transform_function(target, alias).as_sql(self, self.connection)
```
[the remark this answers is not in the document]
Comment (MatthewWilkes) [reply django/db/models/sql/compiler.py]: I'm not sure how this would change to do that, does `as_sql` not take the backend into account? I can't find reference to `as_backend()` in the codebase I'm afraid.

File: django/db/models/sql/query.py (hunk 7 above)
Comment (MatthewWilkes) [reply django/db/models/sql/query.py]: Agreed.

File: django/db/models/sql/query.py (hunk 8 above)
Comment (MatthewWilkes) [reply django/db/models/sql/query.py]: This was done as #8840

File: tests/custom_lookups/tests.py (hunk 9 above)
Comment (MatthewWilkes) [reply tests/custom_lookups/tests.py]: Removed `bilateral` from this transform.

File: django/contrib/postgres/fields/jsonb.py (hunk 10 above)
Comment (MatthewWilkes) [reply django/contrib/postgres/fields/jsonb.py]: I've removed this functionality, it seems the tests pass with it gone, so either something else changed in the interim or I didn't realise I'd changed something elsewhere.

File: django/db/backends/base/operations.py [hunk 12]
```diff
@@ -151,7 +151,7 @@ def distinct_sql(self, fields):
         if fields:
             raise NotImplementedError('DISTINCT ON fields is not supported by this database backend')
         else:
-            return 'DISTINCT'
+            return ['DISTINCT'], []
```
[the remark this answers is not in the document]
Comment (MatthewWilkes) [reply django/db/backends/base/operations.py]: According to the MSSQL backend docs, only one Django version is supported at a time. I think that should make it trivial, but your suggestion works well too. I think this is a safe change.

**Comment by MatthewWilkes** [comment]
> I'd like to see more tests outside contrib/postgres that verify order/values/distinct/f usage. It looks like most of the cases above are actually tested within the json/hstore tests, but that's no good for all other backends.

`distinct on` is a harder one to test, as many backends don't support it. We have `test_transform_order_by` that orders by `age__lastdigit` and the `FLookupTests` class that do values and aggregate. I can write more, but I think these cover the set of functionality so I don't know what extra things to add that would be testing things that aren't currently covered.

> If .objects.values('datefield__month').annotate(sales=Count('pk')) now works (without the additional annotate preceding the values clause) I'd love to see that as a test, and to write up some usage docs showing the ability to write those kinds of queries.

I don't follow this one, I'm afraid?

**Comment by MatthewWilkes** [comment]
I've added a little more in the way of documentation and gone through looking for inconsistencies, but I still have the question from the above comment in October.

File: django/db/models/sql/compiler.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (jarshwah) [reply django/db/models/sql/compiler.py]: `.as_sql()` is the default implementation. `as_{backend.vendor}` is the backend specific implementation *if it exists*.

`compiler.compile` checks if a backend specific implementation exists first and uses that, falling back to `as_sql` if one does not exist.

So, `self.compile(transform_function(..))` I think would be correct.

**Comment by jarshwah** [comment]
> I don't follow this one, I'm afraid?

Now that `field__transform` works in `order_by()`, I'm wondering if it works in `.values()` as well by happenstance.

If you want to group by a transform right now, you have to do this:

```
(
    Model
    .objects
    .annotate(transformed=Transform('field'))
    .values('transformed')
    .annotate(thecount=Count('id'))
)
```

What I want to be able to do is this:

```
(
    Model
    .objects
    .values('field__transform')
    .annotate(thecount=Count('id'))
)
```

I was hoping that the use of `transform_function` from JoinInfo will *just work* for `values()` as well as `order_by()`.

**Comment by MatthewWilkes** [comment]
@jarshwah Regarding `.values()`, that is indedd the case:

Documentation patch is at:
https://github.com/django/django/pull/8528/files#diff-5e313dd49d68a0a867d830a7f442d2a6R538

Tests added are at:
https://github.com/django/django/pull/8528/files#diff-09569af128ce9190c46eaae0d7330a23R1381

File: django/db/models/sql/compiler.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (MatthewWilkes) [reply django/db/models/sql/compiler.py]: Ah, thanks! I'll make this change.

File: django/db/models/sql/compiler.py [hunk 13]
```diff
@@ -622,21 +627,22 @@ def get_distinct(self):
...
-        qn = self.quote_name_unless_alias
-        qn2 = self.connection.ops.quote_name
         result = []
+        params = []
         opts = self.query.get_meta()
 
         for name in self.query.distinct_fields:
             parts = name.split(LOOKUP_SEP)
-            _, targets, alias, joins, path, _ = self._setup_joins(parts, opts, None)
+            field, targets, alias, joins, path, _, transform_function = self._setup_joins(parts, opts, None)
```
[the remark this answers is not in the document]
Comment (MatthewWilkes) [reply django/db/models/sql/compiler.py]: Good catch, it's rebased a few times over the versions, I have no idea if this was a careless copy/paste or a bad rebase, but it can go.

File: django/db/models/sql/query.py [hunk 14]
```diff
@@ -1471,7 +1511,7 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True,
             joins.append(alias)
             if filtered_relation:
                 filtered_relation.path = joins[:]
-        return JoinInfo(final_field, targets, opts, joins, path)
+        return JoinInfo(final_field, targets, opts, joins, path, final_transformer)
```
[the remark this answers is not in the document]
Comment (MatthewWilkes) [reply django/db/models/sql/query.py]: I've added a description of the transform function to the docstring but I didn't include information about the named tuple, perhaps I should have done this as part of #8840 but I think maybe if we think this should reference JoinInfo that can be a small improvement rather than adding to this monster of a ticket?

File: docs/releases/2.1.txt [hunk 15]
```diff
@@ -216,6 +219,9 @@ Database backend API
   feature are changed from :exc:`NotImplementedError` to
   :exc:`django.db.NotSupportedError`.
 
+* The ``DatabaseOperations.distinct_sql()`` method now returns a tuple
+  of SQL and parameters, instead of a SQL string.
```
[the remark this answers is not in the document]
Comment (MatthewWilkes) [reply docs/releases/2.1.txt]: It's worth adding, you're right.

File: django/db/models/sql/query.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (carltongibson) [reply django/db/models/sql/query.py]: I'd rather see the docstring updated in this one: just so it's not forgotten. With the change, the paragraph beginning `Return...` isn't right. 

My worry is if we leave it then it'll be forgotten.

[the remark this answers is not in the document]
Comment (timgraham) [reply django/db/models/sql/query.py]: Please don't suggest doing unrelated cleanups along with bug fix or feature work. We try to avoid that as it confuses things when studying a patch.

File: docs/ref/models/querysets.txt [hunk 16]
```diff
@@ -535,6 +535,18 @@ The ``values()`` method also takes optional keyword arguments,
...
+
+    >>> from django.db.models import CharField
+    >>> from django.db.models.functions import Lower
+    >>> CharField.register_lookup(Lower, 'lower')
+    >>> Blog.objects.values('name__lower')
+    <QuerySet [{'name__lower': 'beatles blog'}]>
+
+The same is true of the built-in lookups, such as ``__date`` or ``__month`` on
+:class:`~django.db.models.DateTimeField`, as described in :doc:`database-functions`.
+
```
[the remark this answers is not in the document]
Comment (MatthewWilkes) [reply docs/ref/models/querysets.txt]: I see what you mean. I'll add a note in the subtleties list but I think I'd be -1 on making a big deal of it in the main narrative as it seems like such an edge case.

File: tests/backends/base/test_operations.py [hunk 17]
```diff
@@ -17,7 +17,7 @@ def setUp(self):
     def test_distinct_on_fields(self):
         msg = 'DISTINCT ON fields is not supported by this database backend'
         with self.assertRaisesMessage(NotSupportedError, msg):
-            self.ops.distinct_sql(['a', 'b'])
+            self.ops.distinct_sql('a', 'b')
```
[the remark this answers is not in the document]
Comment (MatthewWilkes) [reply tests/backends/base/test_operations.py]: Strongly agree.

**Comment by MatthewWilkes** [comment]
Thanks for your comments, @carltongibson! I've pushed a revision that should address the individual line comments. I'm also unsure about the approach for the readme and I haven't added examples in there yet but am open to doing so.

I think that maybe if  we're going to include how to order by a nested JSON key in the documentation it would need to include how to set up the appropriate indexes?

**Comment by timgraham** [comment]
Generally any examples should be included in the documentation and linked to from the release notes rather than included in the release notes.

File: django/db/models/sql/query.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (MatthewWilkes) [reply django/db/models/sql/query.py]: I added the transform function info to the 'Return...' paragraph so that it won't be incorrect, I agree that it's not the clearest docstring in the world an that improving it would be worthwhile, but we already split out the JoinInfo cleanup from this ticket and I don't really want to reintroduce it.

[the remark this answers is not in the document]
Comment (carltongibson) [reply django/db/models/sql/query.py]: Thanks @MatthewWilkes. It was only that sentence/paragraph that needs fixing here, since it's adding the new value that will be unpacked. Any other clean up can wait.

**Comment by timgraham** [comment]
buildbot, test on oracle.

**Comment by MatthewWilkes** [comment]
Should I rebase this to fix the conflict in the release notes, or will the committer do that change? I'm just aware that it'll trigger a dozen builds and remove the confirmation that it passed on oracle.

**Comment by carltongibson** [comment]
@MatthewWilkes I think this was good to go. We should be able to rebase. Thank you!

File: tests/expressions/tests.py [hunk 18]
```diff
@@ -1357,6 +1357,40 @@ def test_raise_empty_expressionlist(self):
             ExpressionList()
 
 
+class FLookupTests(TestCase):
```
Comment (timgraham) [inline tests/expressions/tests.py]: Is the class name correct? I don't see any F expressions -- maybe I misunderstood what "FLookup" means.

File: docs/ref/models/querysets.txt [hunk 19]
```diff
@@ -580,6 +593,23 @@ A few subtleties that are worth mentioning:
...
 
+* Combining transforms and aggregates requires the use of two :meth:`annotate`
+  calls, either explicitly or as keyword arguments to :meth:`values`. As above,
+  if the transform has been registered on the relevant field type the first
+  :meth:`annotate` can be omitted, thus the following examples are equivalent::
+
+    >>> from django.db.models import CharField, Count
+    >>> from django.db.models.functions import Lower
+    >>> CharField.register_lookup(Lower, 'lower')
+    >>> Blog.objects.values('author__name__lower').annotate(entries=Count('entry'))
```
Comment (timgraham) [inline docs/ref/models/querysets.txt]: I think these examples need correcting as `Blog` doesn't have an `author` field.

Comment (MatthewWilkes) [reply docs/ref/models/querysets.txt]: That's a good point. I'd expanded on an example higher in the file, which was fixed in the mainline as #9601. I'll apply the same change here.

File: docs/ref/models/expressions.txt [hunk 20]
```diff
@@ -64,10 +64,11 @@ Some examples
     # Aggregates can contain complex computations also
     Company.objects.annotate(num_offerings=Count(F('products') + F('services')))
 
-    # Expressions can also be used in order_by()
+    # Expressions can also be used in order_by(), either directly
     Company.objects.order_by(Length('name').asc())
     Company.objects.order_by(Length('name').desc())
-
+    # or using the double underscore syntax.
+    Company.objects.order_by('products__count')
```
Comment (timgraham) [inline docs/ref/models/expressions.txt]: Does `__count` need to be registered as a transform for this to work? I guess it's a bit unclear what `products` refers to since the `Company` model isn't defined as far as I see.

File: tests/expressions/tests.py (hunk 18 above)
Comment (MatthewWilkes) [reply tests/expressions/tests.py]: This was from jarshwah's initial hint in the ticket. I agree, your suggested name is better.

File: docs/ref/models/expressions.txt (hunk 20 above)
Comment (MatthewWilkes) [reply docs/ref/models/expressions.txt]: A good point. I'm going to make this use `name__length` instead. It's a bit clearer, as name is a normal field rather than a relationship and we have a similar example in the 1.10 release notes.

**Comment by MatthewWilkes** [comment]
I've pushed a change that I believe addresses the comments from 2 days ago and 20 minutes ago. To be clear, I agree with the changes you've made directly: thanks!
